src/pip/_internal/operations/install/wheel.py

```diff
@@ -656,25 +624,48 @@ def install_wheel(
         )
 
 
-def wheel_version(source_dir):
-    # type: (Optional[str]) -> Tuple[int, ...]
-    """Return the Wheel-Version of an extracted wheel, if possible.
-    Otherwise, raise UnsupportedWheel if we couldn't parse / extract it.
+def wheel_dist_info_dir(source, name):
+    # type: (str, str) -> str
+    """Returns the name of the contained .dist-info directory.
+
+    Raises AssertionError or UnsupportedWheel if not found, >1 found, or
+    it doesn't match the provided name.
     """
-    try:
-        dists = [d for d in pkg_resources.find_on_path(None, source_dir)]
-    except Exception as e:
+    subdirs = os.listdir(source)
+    info_dirs = [s for s in subdirs if s.endswith('.dist-info')]
+
+    if not info_dirs:
+        raise UnsupportedWheel(".dist-info directory not found")
+
+    if len(info_dirs) > 1:
+        raise UnsupportedWheel(
+            "multiple .dist-info directories found: {}".format(
+                ", ".join(info_dirs)
+            )
+        )
+
+    info_dir = info_dirs[0]
+
+    info_dir_name = canonicalize_name(info_dir)
+    canonical_name = canonicalize_name(name)
+    if not info_dir_name.startswith(canonical_name):
```

Comment:
nit (and for a follow-up PR): since we have canonicalized everything we might as well check for an exact match
`if not info_dir_name[:-len('.dist-info')] == canonical_name`
this would catch a messed-up wheel with `canonical_name = foobar` and `info_dir_name = foobar1234.dist-info`.

Reply:
Since the `.dist-info` dir also has version, we would need that too.